src/utils/taint.js

```diff
@@ -0,0 +1,398 @@
+    for (let i = 0; i < len; i++) {
+      let entry = entries[i];
+      if (entry === undefined) continue;
+      this.visitValue(entry);
+    }
+  }
+
+  visitValueFunction(val: FunctionValue): void {
+    this.visitObjectProperties(val);
+
+    if (val instanceof BoundFunctionValue) {
+      this.visitValue(val.$BoundTargetFunction);
+      this.visitValue(val.$BoundThis);
+      for (let boundArg of val.$BoundArguments) this.visitValue(boundArg);
+      return;
+    }
+
+    invariant(
+      !(val instanceof NativeFunctionValue),
+      "all native function values should have already been created outside this pure function"
+    );
+
+    let environment = val.$Environment.parent;
+    while (environment) {
+      let record = environment.environmentRecord;
+      if (record instanceof ObjectEnvironmentRecord) {
+        this.visitValue(record.object);
+        continue;
+      }
+
+      invariant(
+        !(record instanceof GlobalEnvironmentRecord),
+        "we should never reach the global scope because it is never impure in a pure function."
+      );
+      invariant(record instanceof DeclarativeEnvironmentRecord);
+
+      this.visitDeclarativeEnvironmentRecordBinding(record);
+
+      if (record instanceof FunctionEnvironmentRecord) {
+        // If this is a function environment, we visit the function object which if it is
+        // impure will comeback here to visit its parent environment.
+        let fn = record.$FunctionObject;
+        this.visitValue(fn);
+        break;
+      }
+      environment = environment.parent;
+    }
+  }
+
+  visitValueObject(val: ObjectValue): void {
+    let kind = val.getKind();
+    this.visitObjectProperties(val, kind);
+
+    switch (kind) {
+      case "RegExp":
+      case "Number":
+      case "String":
+      case "Boolean":
+      case "ReactElement":
+      case "ArrayBuffer":
+        return;
+      case "Date":
+        let dateValue = val.$DateValue;
+        invariant(dateValue !== undefined);
+        this.visitValue(dateValue);
+        return;
+      case "Float32Array":
+      case "Float64Array":
+      case "Int8Array":
+      case "Int16Array":
+      case "Int32Array":
+      case "Uint8Array":
+      case "Uint16Array":
+      case "Uint32Array":
+      case "Uint8ClampedArray":
+      case "DataView":
+        let buf = val.$ViewedArrayBuffer;
+        invariant(buf !== undefined);
+        this.visitValue(buf);
+        return;
+      case "Map":
+      case "WeakMap":
+        this.visitValueMap(val);
+        return;
+      case "Set":
+      case "WeakSet":
+        this.visitValueSet(val);
+        return;
+      default:
+        invariant(kind === "Object", `Object of kind ${kind} is not supported in calls to abstract functions.`);
+        invariant(
+          this.$ParameterMap === undefined,
+          `Arguments object is not supported in calls to abstract functions.`
+        );
+        return;
+    }
+  }
+
+  visitValueSymbol(val: SymbolValue): void {
```

Comment:
This is never called.

Reply:
Yea. Not sure about this. The description could be abstract, which could have objects as their arguments, but since there is no way for code to access those, I think it is safe to never visit symbols. Will remove this path.